bodyparsers/msgpack/README.md

```diff
@@ -0,0 +1,185 @@
+# @loopback/bodyparser-msgpack
+
+This module extends LoopBack with the ability to receive
+[MessagePack](https://msgpack.org/) requests and transparently convert it to a
+regular JavaScript object. It provides a BodyParser implementation and a
+component to register it.
+
+## Stability: ⚠️Experimental⚠️
+
+> Experimental packages provide early access to advanced or experimental
+> functionality to get community feedback. Such modules are published to npm
+> using `0.x.y` versions. Their APIs and functionality may be subject to
+> breaking changes in future releases.
+
+## Installation
+
+```sh
+npm i @loopback/bodyparser-msgpack --save
+```
+
+## Usage
+
+The component should be loaded in the constructor of your custom Application
+class.
+
+Start by importing the component class:
+
+```ts
+import {MsgPackBodyParserComponent} from '@loopback/bodyparser-msgpack';
+```
+
+In the constructor, add the component to your application:
+
+```ts
+this.component(MsgPackBodyParserComponent);
+```
+
+The body parser will accept requests with the following MIME type
+(`Content-Type`) blobs:
+
+- `application/msgpack`
+- `application/x-msgpack`
+- `application/*+msgpack`
+
+### Accepting MessagePack Requests
+
+To accept MessagePack requests in a controller, amend the OpenAPI decorator to
+include the MIME type as a possible requset body.
+
+For example, to update the Todo controller to accept MessagePack:
+
+```typescript
+import {post, getModelSchemaRef, requestBody} from '@loopback/rest';
+
+class TodoController {
+  // Omitted constructor for bevity
+
+  @post('/todos')
+  async create(
+    @requestBody({
+      content: {
+        // Change existing or append a new requset body accepted MIME type
+        'application/msgpack': {
+          schema: getModelSchemaRef(Todo, {
+            title: 'NewTodo',
+            exclude: ['id'],
+          }),
+        },
+      },
+    })
+    todo: // Keep the response body object type, since the body parser transparently
+    // converts it into a JavaScript object.
+    Omit<Todo, 'id'>,
+
+    // For bevity, the function does not return anything. See
+    // 'Returning MessagePack Requests' below.
+  ): void {
+    this.todoRepository.create(todo);
+  }
+}
+```
+
+The MessagePack request payload will be transparently converted into a
+JavaScript object and validated against the JSON Schema.
+
+### Returning MessagePack Requests
+
+{% include note.html content="The body parser will not convert responses into `application/msgpack`. This needs to be done manually." %}
```

Comment:
A few lines down, the README (line 172) references #436. Would that issue cover this feature?

Reply:
Most likely not. As I understand  #436, it will allow controllers to returns something like this:

```ts
class MyController {
  @get('/')
  async index() {
    return {
      statusCode: 200,
      headers: {'content-type': 'application/msgpack'},
      body: msgpack.encode({hello: 'world'}),
    };
  }
}
```

While this is simpler than the code you have in your example, it's still quite involved and it does not support content-type negotiation based on `Accepts` request header.

The solution I am looking for:

- The extension tells LoopBack that `application/msgpack` is a supported response type now and provides a code for converting JS objects into response bodies.
- Runtime (`@loopback/rest`) parses `Accepts` header, finds the intersection between the content types accepted by the client and supported by the server (us), and decided what type to use for the response. Then it calls the appropriate serializer for the selected content type.
- A LoopBack user registers the msgpack extension. No other changes in controllers are necessary.

```ts
class MyController {
  @get('/')
  async index() {
    return {hello: 'world'}'
  }
}
```

Another missing piece is how to include msgpack content type in OpenAPI metadata describing the supported response formats. AFAIK, our `@oas.response` decorator always describes `application/json` only.

https://github.com/strongloop/loopback-next/blob/417245aa28df0b34f62921438bdedb8fd15cadda/packages/openapi-v3/src/decorators/response.decorator.ts#L29

Supporting multiple content-type with proper negotiation is not a trivial task 🤷🏻

Reply:
Opened #6275

Reply:
> Opened #6275

Thank you!